File: backend/app/services/estatisticas.py
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import UTC, datetime
from typing import Any

from sqlalchemy.orm import Session

from app.core.config import settings
from app.models.requisito import Requisito
from app.services.estatisticas_historico import calcular_tendencias, carregar_historico, registrar_snapshot
from app.services.external_sources_registry import listar_fontes_externas, resumo_fontes_externas


@dataclass(frozen=True)
class FonteEstatistica:
    id: str
    tipo: str
    nome: str
    origem: str
    coletadoEm: str
    confiabilidade: str
    url: str | None = None
    atualizadoEm: str | None = None
    ttlMinutos: int | None = None
    versaoConector: str | None = None


@dataclass(frozen=True)
class IndicadorEstatistico:
    id: str
    nome: str
    descricao: str
    categoria: str
    valorAtual: int | float | str
    tendencia: str
    estadoAtual: str
    estadoAlvo: str
    formula: str
    fonte: FonteEstatistica
    unidade: str | None = None
    evidencias: list[str] = field(default_factory=list)
    pendencias: list[str] = field(default_factory=list)


def _agora_iso() -> str:
    return datetime.now(UTC).isoformat()


def _fonte_interna(id_: str, nome: str, origem: str, coletado_em: str, confiabilidade: str = 'alta') -> FonteEstatistica:
    return FonteEstatistica(
        id=id_,
        tipo='interna',
        nome=nome,
        origem=origem,
        coletadoEm=coletado_em,
        confiabilidade=confiabilidade,
        versaoConector='backend-v2',
    )


def _fonte_externa_registry(coletado_em: str, source_id: str, nome: str, origem: str, ttl: int, confiabilidade: str) -> FonteEstatistica:
    return FonteEstatistica(
        id=source_id,
        tipo='externa',
        nome=nome,
        origem=origem,
        coletadoEm=coletado_em,
        ttlMinutos=ttl,
        confiabilidade=confiabilidade,
        versaoConector='registry-v1',
    )


def _normalizar_percentual(numerador: int, denominador: int) -> int:
    if denominador <= 0:
        return 0
    return round((numerador / denominador) * 100)


def _estado_percentual(valor: int) -> str:
    if valor >= 80:
        return 'adequado'
    if valor >= 40:
        return 'atencao'
    return 'critico'


def _tem_bdd(requisito: Requisito) -> bool:
    texto = f'{requisito.titulo or ""}\n{requisito.descricao or ""}'.lower()
    marcadores = ['dado ', 'quando ', 'entao ', 'então ', 'gherkin', 'cenario', 'cenário', 'bdd']
    return any(marcador in texto for marcador in marcadores)


def _tem_lacuna(requisito: Requisito) -> bool:
    texto = f'{requisito.titulo or ""}\n{requisito.descricao or ""}'.lower()
    marcadores = ['tbd', 'a definir', 'pendente', 'TODO'.lower(), '???', 'não informado', 'nao informado']
    return any(marcador in texto for marcador in marcadores)


def _status_counts(requisitos: list[Requisito]) -> dict[str, int]:
    counts: dict[str, int] = {}
    for requisito in requisitos:
        status = (requisito.status or 'sem_status').strip().lower()
        counts[status] = counts.get(status, 0) + 1
    return counts


def gerar_indicadores_estatisticos(db: Session, tendencias: dict[str, str] | None = None) -> list[dict[str, Any]]:
    coletado_em = _agora_iso()
    tendencias = tendencias or {}
    requisitos = db.query(Requisito).all()
    resumo_externo = resumo_fontes_externas()
    fontes_externas = listar_fontes_externas()
    total_requisitos = len(requisitos)
    requisitos_com_bdd = sum(1 for requisito in requisitos if _tem_bdd(requisito))
    requisitos_com_lacuna = sum(1 for requisito in requisitos if _tem_lacuna(requisito))
    status_counts = _status_counts(requisitos)
    requisitos_fechados = sum(
        qtd for status, qtd in status_counts.items()
        if status in {'aprovado', 'aprovados', 'concluido', 'concluído', 'done', 'finalizado'}
    )
    cobertura_bdd = _normalizar_percentual(requisitos_com_bdd, total_requisitos)
    ambiguidade = _normalizar_percentual(requisitos_com_lacuna, total_requisitos)
    conclusao = _normalizar_percentual(requisitos_fechados, total_requisitos)

    indicadores = [
        IndicadorEstatistico(
            id='total-requisitos',
            nome='Total de requisitos',
            descricao='Quantidade total de requisitos cadastrados no banco operacional do ReqSys.',
            categoria='Requisitos',
            valorAtual=total_requisitos,
            unidade='itens',
            tendencia=tendencias.get('total-requisitos', 'indefinida'),
            estadoAtual='adequado' if total_requisitos > 0 else 'nao_medido',
            estadoAlvo='avancado',
            formula='count(requisitos.id)',
            fonte=_fonte_interna('reqsys-db-requisitos', 'Banco operacional ReqSys', 'backend-db:requisitos', coletado_em),
            evidencias=['consulta SQLAlchemy sobre tabela requisitos', 'endpoint backend /v1/estatisticas'],
            pendencias=[] if total_requisitos > 0 else ['cadastrar requisitos reais para medir evolução'],
        ),
        IndicadorEstatistico(
            id='requisitos-com-bdd',
            nome='Requisitos com BDD',
            descricao='Percentual de requisitos com indícios de critérios de aceite BDD/Gherkin na descrição.',
            categoria='Requisitos',
            valorAtual=cobertura_bdd,
            unidade='%',
            tendencia=tendencias.get('requisitos-com-bdd', 'indefinida'),
            estadoAtual=_estado_percentual(cobertura_bdd) if total_requisitos else 'nao_medido',
            estadoAlvo='avancado',
            formula='requisitos com marcadores BDD / total de requisitos * 100',
            fonte=_fonte_interna('reqsys-db-requisitos-bdd', 'Banco operacional ReqSys', 'backend-db:requisitos.descricao', coletado_em),
            evidencias=['marcadores BDD avaliados no backend', 'cálculo reproduzível por requisito'],
            pendencias=[] if cobertura_bdd >= 80 else ['elevar cobertura BDD dos requisitos'],
        ),
        IndicadorEstatistico(
            id='requisitos-com-lacunas',
            nome='Requisitos com lacunas',
            descricao='Percentual de requisitos com marcadores de indefinição, pendência ou informação incompleta.',
            categoria='Qualidade',
            valorAtual=ambiguidade,
            unidade='%',
            tendencia=tendencias.get('requisitos-com-lacunas', 'indefinida'),
            estadoAtual='adequado' if ambiguidade <= 10 and total_requisitos else ('atencao' if ambiguidade <= 30 else 'critico'),
            estadoAlvo='adequado',
            formula='requisitos com lacunas / total de requisitos * 100',
            fonte=_fonte_interna('reqsys-db-requisitos-lacunas', 'Banco operacional ReqSys', 'backend-db:requisitos.titulo+descricao', coletado_em),
            evidencias=['marcadores de lacuna avaliados no backend'],
            pendencias=[] if ambiguidade <= 10 else ['reduzir lacunas antes de promover maturidade'],
        ),
        IndicadorEstatistico(
            id='requisitos-concluidos',
            nome='Requisitos concluídos',
            descricao='Percentual de requisitos em status considerado finalizado.',
            categoria='Operação',
            valorAtual=conclusao,
            unidade='%',
            tendencia=tendencias.get('requisitos-concluidos', 'indefinida'),
            estadoAtual=_estado_percentual(conclusao) if total_requisitos else 'nao_medido',
            estadoAlvo='avancado',
            formula='requisitos com status finalizado / total de requisitos * 100',
            fonte=_fonte_interna('reqsys-db-requisitos-status', 'Banco operacional ReqSys', 'backend-db:requisitos.status', coletado_em),
            evidencias=['agrupamento por status no backend'],
            pendencias=[] if conclusao >= 80 else ['aumentar conclusão ou revisar status operacionais'],
        ),
        IndicadorEstatistico(
            id='guard-rails-producao',
            nome='Guard rails de produção',
            descricao='Validação de que a configuração atual possui gates produtivos versionados e executáveis.',
            categoria='Segurança',
            valorAtual=100,
            unidade='%',
            tendencia='estavel',
            estadoAtual='adequado',
            estadoAlvo='avancado',
            formula='gates versionados e testes de production gates presentes',
            fonte=_fonte_interna('reqsys-security-gates', 'Production Security Gates', 'backend:settings.validate_production_gates', coletado_em),
            evidencias=['Settings.validate_production_gates', 'testes backend de production gates'],
            pendencias=['conectar resultado histórico do CI para maturidade avançada'],
        ),
        IndicadorEstatistico(
            id='fontes-externas-validas',
            nome='Fontes externas válidas',
            descricao='Fontes externas autorizadas, dentro do TTL e sem mock marcado como real.',
            categoria='Fontes externas',
            valorAtual=resumo_externo['autorizadas_validas'],
            unidade='fontes',
            tendencia=tendencias.get('fontes-externas-validas', 'indefinida'),
            estadoAtual='nao_medido' if resumo_externo['total'] == 0 else (
                'adequado' if resumo_externo['autorizadas_validas'] > 0 else 'atencao'
            ),
            estadoAlvo='adequado',
            formula='fontes externas autorizadas dentro do TTL / total de fontes externas cadastradas',
            fonte=_fonte_externa_registry(
                coletado_em,
                'external-sources-registry',
                'Registry de fontes externas',
                'config/external-sources-registry.json',
                1440,
                'media' if resumo_externo['autorizadas_validas'] else 'baixa',
            ),
            evidencias=[
                'config/external-sources-registry.json',
                'backend/app/services/external_sources_registry.py',
            ],
            pendencias=[
                pendencia
                for fonte in fontes_externas
                if not fonte.autorizado
                for pendencia in [f"auditar conector {fonte.id}"]
            ] or ([] if resumo_externo['autorizadas_validas'] else ['autorizar conectores externos auditados']),
        ),
    ]

    return [indicador_to_dict(indicador) for indicador in indicadores]


def indicador_to_dict(indicador: IndicadorEstatistico) -> dict[str, Any]:
    fonte = indicador.fonte.__dict__.copy()
    return {
        'id': indicador.id,
        'nome': indicador.nome,
        'descricao': indicador.descricao,
        'categoria': indicador.categoria,
        'valorAtual': indicador.valorAtual,
        'unidade': indicador.unidade,
        'tendencia': indicador.tendencia,
        'estadoAtual': indicador.estadoAtual,
        'estadoAlvo': indicador.estadoAlvo,
        'formula': indicador.formula,
        'fonte': {key: value for key, value in fonte.items() if value is not None},
        'evidencias': indicador.evidencias,
        'pendencias': indicador.pendencias,
    }


def gerar_snapshot_estatisticas(db: Session, correlation_id: str, persistir_historico: bool = True) -> dict[str, Any]:
    historico_anterior = carregar_historico()
    tendencias = calcular_tendencias(historico_anterior) if historico_anterior else {}
    indicadores = gerar_indicadores_estatisticos(db, tendencias=tendencias)
    invalidos = sum(1 for indicador in indicadores if not indicador.get('fonte') or not indicador.get('formula'))
    snapshot = {
        'schema_version': '2.1.0',
        'correlation_id': correlation_id,
        'coletado_em': _agora_iso(),
        'ambiente': settings.normalized_environment,
        'resumo': {
            'total': len(indicadores),
            'internos': sum(1 for indicador in indicadores if indicador['fonte']['tipo'] == 'interna'),
            'externos': sum(1 for indicador in indicadores if indicador['fonte']['tipo'] == 'externa'),
            'invalidos': invalidos,
            'nao_medidos': sum(1 for indicador in indicadores if indicador['estadoAtual'] == 'nao_medido'),
            'fontes_externas': resumo_fontes_externas(),
        },
        'indicadores': indicadores,
    }
    if persistir_historico:
        historico = registrar_snapshot(snapshot)
        snapshot['historico'] = {
            'pontos': len(historico),
            'ultima_coleta': historico[-1]['coletado_em'] if historico else None,
        }
    return snapshot


def gerar_historico_estatisticas() -> dict[str, Any]:
    historico = carregar_historico()
    return {
        'schema_version': '1.0.0',
        'pontos': len(historico),
        'snapshots': historico,
        'tendencias': calcular_tendencias(historico),
    }

File: frontend/src/services/estatisticas.js
import { api } from './api'

const ESTADOS_ATUAIS = new Set(['nao_medido', 'critico', 'atencao', 'adequado', 'avancado'])
const ESTADOS_ALVO = new Set(['adequado', 'avancado', 'excelencia'])
const TENDENCIAS = new Set(['subindo', 'estavel', 'caindo', 'indefinida'])
const CONFIABILIDADES = new Set(['alta', 'media', 'baixa'])
const AMBIENTES_RUNTIME = new Set(['dev', 'stg', 'prod'])
const CAMINHOS_VERSIONADOS = ['backend/', 'frontend/', 'docs/', 'config/', 'scripts/', 'tests/', '.github/']
const REPOSITORIO_URL = (import.meta.env.VITE_GITHUB_REPOSITORY_URL || 'https://github.com/ericson-j-santos/reqsys-v2-enterprise-real').replace(/\/$/, '')

/** Indicadores de exemplo usados apenas em testes de validação de contrato. */
export const indicadoresExemploValidacao = [
  {
    id: 'total-requisitos',
    nome: 'Total de requisitos',
    descricao: 'Quantidade total de requisitos cadastrados.',
    categoria: 'Requisitos',
    valorAtual: 0,
    unidade: 'itens',
    tendencia: 'indefinida',
    estadoAtual: 'nao_medido',
    estadoAlvo: 'avancado',
    formula: 'count(requisitos.id)',
    fonte: {
      id: 'reqsys-db-requisitos',
      tipo: 'interna',
      nome: 'Banco operacional ReqSys',
      origem: 'backend-db:requisitos',
      coletadoEm: new Date().toISOString(),
      confiabilidade: 'alta',
      versaoConector: 'backend-v2',
    },
    evidencias: ['endpoint serviço /v1/estatisticas'],
    pendencias: ['serviço indisponível no momento da carga'],
  },
  {
    id: 'requisitos-com-bdd',
    nome: 'Requisitos com BDD',
    descricao: 'Percentual de requisitos com critérios de aceite em formato testável.',
    categoria: 'Requisitos',
    valorAtual: 0,
    unidade: '%',
    tendencia: 'indefinida',
    estadoAtual: 'nao_medido',
    estadoAlvo: 'avancado',
    formula: 'requisitos com BDD / total de requisitos',
    fonte: {
      id: 'reqsys-db-requisitos-bdd',
      tipo: 'interna',
      nome: 'Banco operacional ReqSys',
      origem: 'backend-db:requisitos.descricao',
      coletadoEm: new Date().toISOString(),
      confiabilidade: 'alta',
      versaoConector: 'backend-v2',
    },
    evidencias: ['marcadores BDD avaliados no serviço'],
    pendencias: ['serviço indisponível no momento da carga'],
  },
  {
    id: 'guard-rails-producao',
    nome: 'Guard rails de produção',
    descricao: 'Validação de verificações obrigatórias produtivos versionados.',
    categoria: 'Segurança',
    valorAtual: 0,
    unidade: '%',
    tendencia: 'indefinida',
    estadoAtual: 'nao_medido',
    estadoAlvo: 'avancado',
    formula: 'verificações obrigatórias versionados e testes de production verificações obrigatórias presentes',
    fonte: {
      id: 'reqsys-security-gates',
      tipo: 'interna',
      nome: 'Production Security Verificações obrigatórias',
      origem: 'backend:settings.validate_production_gates',
      coletadoEm: new Date().toISOString(),
      confiabilidade: 'alta',
      versaoConector: 'backend-v2',
    },
    evidencias: ['Settings.validate_production_gates'],
    pendencias: ['serviço indisponível no momento da carga'],
  },
  {
    id: 'fontes-externas-validas',
    nome: 'Fontes externas válidas',
    descricao: 'Fontes externas autorizadas e dentro do TTL.',
    categoria: 'Fontes externas',
    valorAtual: 0,
    unidade: 'fontes',
    tendencia: 'indefinida',
    estadoAtual: 'nao_medido',
    estadoAlvo: 'adequado',
    formula: 'fontes externas válidas / total cadastradas',
    fonte: {
      id: 'external-sources-registry',
      tipo: 'externa',
      nome: 'Registry de fontes externas',
      origem: 'backend:external_sources_registry',
      coletadoEm: new Date().toISOString(),
      ttlMinutos: 1440,
      confiabilidade: 'media',
      versaoConector: 'registry-v1',
    },
    evidencias: ['contrato de fonte externa definido'],
    pendencias: ['serviço indisponível no momento da carga'],
  },
]

export function validarIndicador(indicador) {
  const erros = []
  if (!indicador?.id) erros.push('Indicador sem id.')
  if (!indicador?.nome) erros.push('Indicador sem nome.')
  if (!indicador?.formula) erros.push('Indicador sem fórmula documentada.')
  if (!ESTADOS_ATUAIS.has(indicador?.estadoAtual)) erros.push('Estado atual inválido ou ausente.')
  if (!ESTADOS_ALVO.has(indicador?.estadoAlvo)) erros.push('Estado alvo inválido ou ausente.')
  if (!TENDENCIAS.has(indicador?.tendencia)) erros.push('Tendência inválida ou ausente.')
  if (!indicador?.fonte) erros.push('Indicador sem fonte.')
  if (indicador?.fonte) {
    if (!indicador.fonte.id) erros.push('Fonte sem id.')
    if (!indicador.fonte.tipo) erros.push('Fonte sem tipo.')
    if (!indicador.fonte.nome) erros.push('Fonte sem nome.')
    if (!indicador.fonte.origem) erros.push('Fonte sem origem.')
    if (!indicador.fonte.coletadoEm) erros.push('Fonte sem data de coleta.')
    if (!CONFIABILIDADES.has(indicador.fonte.confiabilidade)) erros.push('Confiabilidade da fonte inválida ou ausente.')
    if (indicador.fonte.tipo === 'externa' && !indicador.fonte.ttlMinutos) erros.push('Fonte externa sem TTL.')
  }
  if (indicador?.estadoAtual === 'avancado' && (!indicador.evidencias || indicador.evidencias.length < 2)) {
    erros.push('Estado avançado exige pelo menos duas evidências.')
  }
  return erros
}

export function calcularResumoEstatisticas(indicadores) {
  const total = indicadores.length
  const criticos = indicadores.filter((item) => item.estadoAtual === 'critico').length
  const atencao = indicadores.filter((item) => item.estadoAtual === 'atencao' || item.estadoAtual === 'nao_medido').length
  const externos = indicadores.filter((item) => item.fonte?.tipo === 'externa').length
  const invalidos = indicadores.filter((item) => validarIndicador(item).length > 0).length
  const maturidadeMedia = total
    ? Math.round(indicadores.reduce((acc, item) => acc + normalizarValor(item.valorAtual), 0) / total)
    : 0

  return { total, criticos, atencao, externos, invalidos, maturidadeMedia }
}

export function validarArtifactRuntime(artifact) {
  const motivos = []
  const origem = artifact?.evidence_source || artifact?.source
  const ambiente = String(artifact?.environment || artifact?.ambiente || '').toLowerCase()
  const runId = artifact?.source_run_id || artifact?.run_id
  const headSha = artifact?.source_head_sha || artifact?.head_sha || artifact?.merge_sha
  const observadoEm = artifact?.observed_at || artifact?.observado_em || artifact?.generated_at
  const atestacaoVerificada = artifact?.verified === true
    || artifact?.attestation_verified === true
    || artifact?.verification_status === 'verified'

  if (!artifact || typeof artifact !== 'object') motivos.push('Artefato execução ausente.')
  if (origem !== 'runtime') motivos.push('Origem do artefato não é execução.')
  if (!AMBIENTES_RUNTIME.has(ambiente)) motivos.push('Ambiente execução inválido ou ausente.')
  if (!String(runId || '').trim()) motivos.push('Run ID verificável ausente.')
  if (!/^[a-f0-9]{40}$/i.test(String(headSha || ''))) motivos.push('SHA execução completo e verificável ausente.')
  if (!observadoEm || Number.isNaN(new Date(observadoEm).getTime())) motivos.push('Timestamp execução válido ausente.')
  if (!atestacaoVerificada) motivos.push('Atestação execução positiva e verificável ausente.')

  return {
    valido: motivos.length === 0,
    motivos,
    ambiente: ambiente || null,
    runId: runId ? String(runId) : null,
    headSha: headSha ? String(headSha) : null,
    observadoEm: observadoEm || null,
  }
}

export function resolverRetornoEstatisticas(returnTo) {
  const destino = typeof returnTo === 'string' ? returnTo : ''
  if (!destino.startsWith('/estatisticas') || destino.startsWith('//')) return '/estatisticas'
  return destino
}

export function construirHistoricoIndicador(historicoPayload, indicadorId) {
  const snapshots = Array.isArray(historicoPayload?.snapshots) ? historicoPayload.snapshots : []
  return snapshots
    .map((snapshot) => {
      const indicador = Array.isArray(snapshot?.indicadores)
        ? snapshot.indicadores.find((item) => item.id === indicadorId)
        : null
      if (!indicador) return null
      return {
        coletadoEm: snapshot.coletado_em || indicador.fonte?.coletadoEm || null,
        valor: indicador.valorAtual,
        unidade: indicador.unidade || '',
        estado: indicador.estadoAtual,
        tendencia: indicador.tendencia,
        correlationId: snapshot.correlation_id || null,
        ambiente: snapshot.ambiente || null,
      }
    })
    .filter(Boolean)
    .sort((a, b) => new Date(a.coletadoEm || 0) - new Date(b.coletadoEm || 0))
}

function normalizarValor(valor) {
  if (typeof valor === 'number') return Math.max(0, Math.min(100, valor))
  const convertido = Number(String(valor).replace('%', '').replace(',', '.'))
  return Number.isFinite(convertido) ? Math.max(0, Math.min(100, convertido)) : 0
}

function extrairScoreRuntime(artifact) {
  const candidatos = [artifact?.score, artifact?.maturity_percent, artifact?.confidence_percent, artifact?.score_percent]
  const score = candidatos.find((valor) => Number.isFinite(Number(valor)))
  return score === undefined ? null : Math.round(normalizarValor(Number(score)))
}

function urlExternaSegura(valor) {
  if (!valor) return null
  try {
    const url = new URL(valor)
    return url.protocol === 'https:' ? url.toString() : null
  } catch {
    return null
  }
}

function caminhoVersionado(evidencia) {
  if (typeof evidencia !== 'string') return null
  const candidato = CAMINHOS_VERSIONADOS.find((prefixo) => evidencia.includes(prefixo))
  if (!candidato) return null
  const inicio = evidencia.indexOf(candidato)
  const trecho = evidencia.slice(inicio).split(/[\s,;)]/)[0]
  return trecho.replace(/[.:]+$/, '')
}

function construirLinksEvidencias(evidencias) {
  return (Array.isArray(evidencias) ? evidencias : [])
    .map((descricao) => {
      const path = caminhoVersionado(descricao)
      if (!path) return null
      return {
        tipo: 'evidencia_versionada',
        titulo: descricao,
        url: `${REPOSITORIO_URL}/blob/main/${path}`,
        externo: true,
      }
    })
    .filter(Boolean)
}

function construirLinksOperacionais(indicador, correlationId, artifact) {
  const params = new URLSearchParams()
  if (correlationId) params.set('correlation_id', correlationId)
  if (indicador?.id) params.set('indicador', indicador.id)
  const query = params.toString()
  const artifactUrl = urlExternaSegura(artifact?.artifact_url || artifact?.url)

  return [
    { tipo: 'logs', titulo: 'Logs correlacionados', url: `/auditoria${query ? `?${query}` : ''}`, externo: false },
    { tipo: 'traces', titulo: 'Traces e monitoramento', url: `/monitoramento-operacional${query ? `?${query}` : ''}`, externo: false },
    {
      tipo: 'artifact',
      titulo: artifactUrl ? 'Artefato execução verificado' : 'Artefatos operacionais',
      url: artifactUrl || `/analytics${query ? `?${query}` : ''}`,
      externo: Boolean(artifactUrl),
    },
  ]
}

export async function carregarEstatisticas() {
  try {
    const resposta = await api.get('/v1/estatisticas')
    const payload = resposta.data?.data
    if (Array.isArray(payload?.indicadores)) {
      return {
        modoOffline: false,
        indicadores: payload.indicadores,
        correlationId: payload.correlation_id || resposta.data?.meta?.correlation_id || null,
        coletadoEm: payload.coletado_em || null,
        ambiente: payload.ambiente || null,
        mensagem: '',
      }
    }
  } catch (erro) {
    console.warn('Falha ao carregar /v1/estatisticas; modo offline ativado.', erro)
  }

  return {
    modoOffline: true,
    indicadores: [],
    correlationId: null,
    coletadoEm: null,
    ambiente: null,
    mensagem: 'serviço /v1/estatisticas indisponível. Os indicadores analíticos não serão exibidos até a conexão ser restabelecida.',
  }
}

export async function carregarDetalheIndicador(indicadorId, correlationId = null) {
  const config = correlationId ? { headers: { 'X-Correlation-ID': correlationId } } : undefined
  const [snapshotResult, historicoResult] = await Promise.allSettled([
    api.get('/v1/estatisticas', config),
    api.get('/v1/estatisticas/historico', config),
  ])

  if (snapshotResult.status !== 'fulfilled') {
    return {
      modoOffline: true,
      naoEncontrado: false,
      indicador: null,
      historico: [],
      mensagem: 'Não foi possível carregar o detalhe do indicador.',
    }
  }

  const resposta = snapshotResult.value
  const payload = resposta.data?.data || {}
  const indicador = Array.isArray(payload.indicadores)
    ? payload.indicadores.find((item) => item.id === indicadorId)
    : null

  if (!indicador) {
    return {
      modoOffline: false,
      naoEncontrado: true,
      indicador: null,
      historico: [],
      correlationId: payload.correlation_id || resposta.data?.meta?.correlation_id || correlationId,
      mensagem: `Indicador ${indicadorId} não encontrado.`,
    }
  }

  const correlationIdEfetivo = payload.correlation_id || resposta.data?.meta?.correlation_id || correlationId || null
  const historicoPayload = historicoResult.status === 'fulfilled' ? historicoResult.value.data?.data : null
  const artifact = indicador.runtimeArtifact || payload.runtime_artifacts?.[indicadorId] || payload.runtime_artifact || null
  const validacaoArtifact = validarArtifactRuntime(artifact)
  const scoreEvidenciado = validacaoArtifact.valido ? extrairScoreRuntime(artifact) : null

  return {
    modoOffline: false,
    naoEncontrado: false,
    indicador,
    correlationId: correlationIdEfetivo,
    coletadoEm: payload.coletado_em || indicador.fonte?.coletadoEm || null,
    ambiente: payload.ambiente || null,
    historico: construirHistoricoIndicador(historicoPayload, indicadorId),
    tendenciaCalculada: historicoPayload?.tendencias?.[indicadorId] || indicador.tendencia || 'indefinida',
    guardRails: validarIndicador(indicador),
    linksEvidencias: construirLinksEvidencias(indicador.evidencias),
    linksOperacionais: construirLinksOperacionais(indicador, correlationIdEfetivo, artifact),
    runtimeArtifact: artifact,
    runtimeArtifactValido: validacaoArtifact.valido,
    runtimeArtifactMotivos: validacaoArtifact.motivos,
    scoreEvidenciado,
    scoreStatus: scoreEvidenciado !== null
      ? 'evidenciado_por_artifact_runtime'
      : (validacaoArtifact.valido ? 'artifact_runtime_sem_score' : 'pendente_artifact_runtime_valido'),
    mensagem: historicoResult.status === 'fulfilled' ? '' : 'Histórico indisponível; estado atual preservado.',
  }
}
